libs/wingsdk/src/core/counter.ts

```diff
@@ -0,0 +1,59 @@
+import { IConstruct } from "constructs";
+import { App } from "./app";
+
+const COUNTERS_SYMBOL = Symbol.for("@winglang/sdk.core.Counter");
+
+/**
+ * A class that generates tree-unique sequenced IDs for constructs.
+ */
+export class Counters {
+  /**
+   * Returns the singleton instance of the `Counters` class.
+   */
+  public static of(construct: IConstruct): Counters {
+    let connections = (App.of(construct) as any)[COUNTERS_SYMBOL];
+
+    if (!connections) {
+      connections = new Counters();
+      (App.of(construct) as any)[COUNTERS_SYMBOL] = connections;
+    }
+
+    return connections;
+  }
+
+  /**
+   * Generate a unique ID for the given scope and id. The newly generated ID is
+   * guaranteed to be unique within the given scope.
+   * It will have the form '<id><n>', where '<id>' is the given id and '<n>' is a
+   * increasing sequence of integers starting from '0'.
+   *
+   * @param scope The scope for which this id is generated. Also used to get the root singleton
+   */
+  public static createId(scope: IConstruct, id: string): string {
+    const counters = Counters.of(scope);
+
+    return counters.createId(scope, id);
+  }
+
+  private readonly _counters: Record<string, number> = {};
+  private constructor() {}
+
+  /**
+   * Generate a unique ID for the given scope and id. The newly generated ID is
+   * guaranteed to be unique within the given scope.
+   * It will have the form '<id><n>', where '<id>' is the given id and '<n>' is a
+   * increasing sequence of integers starting from '0'.
+   *
+   * @param scope The scope for which this id is generated.
+   */
+  public createId(scope: IConstruct, id: string): string {
+    const key = `${scope.node.addr}|${id}`;
+    const counter = this._counters[key];
+    if (counter !== undefined) {
+      return `${id}${this._counters[key]++}`;
+    } else {
+      this._counters[key] = 1;
+      return `${id}0`;
+    }
```

Comment:
```suggestion
    this._counters[key] = this._counters[key] ?? 0;
    return `${id}_${this._counters[key]++}`;
```

(also add `_` between prefix and counter.

Reply:
Not sure I agree with having `_`. It would be very surprising to use this for non-alpha characters, so you already have a natural separator. As a result, having _ just feels like adding noise (especially because we all use `-` in IDs)

Reply:
ok, no problem